src/app/pages/checkout/checkout.component.ts

```diff
@@ -1,15 +1,129 @@
+    public stars: number[] = [1, 2, 3, 4, 5];
+    private unsubscribeAll = new Subject();
+
+    public options: string[] = CITIES;
+    public filteredOptions: Observable<string[]>;
+
+    public checkoutForm: FormGroup = new FormGroup({
+        firstName: new FormControl('', [Validators.required, Validators.minLength(3)]),
+        lastName: new FormControl(''),
+        email: new FormControl('', [Validators.required, Validators.email]),
+        phone: new FormControl('', [Validators.required, Validators.minLength(10)]),
+        address: new FormControl(''),
+        city: new FormControl(''),
+        country: new FormControl(''),
+        zip: new FormControl('', [Validators.required]),
+        additionalInfo: new FormControl(''),
+        promo: new FormControl(''),
+        mailing: new FormControl(false, [Validators.requiredTrue]),
+        terms: new FormControl(false, [Validators.requiredTrue]),
+        goods: new FormArray([], [Validators.required])
+    });
+
+    constructor(private cartService: CartService) {
+    }
+
+    ngOnInit(): void {
+        this.getGoods();
+        this.getSum();
+        this.getFilteredOptions();
+        this.patchGoods();
+    }
+
+    private patchGoods(): void {
+        const control = this.checkoutForm.get('goods') as FormArray;
+        control.clear();
+        this.goods.forEach(item => {
+            control.push(new FormControl({
+                count: item.count,
+                good: item.good
+            }));
+        });
+    }
+
+    private getFilteredOptions(): void {
+        this.filteredOptions = this.checkoutForm.get('city').valueChanges.pipe(
+            startWith(''),
+            map(value => this.filterOptions(value))
+        );
+    }
+
+    private filterOptions(value: string): string[] {
+        const filterValue = value.toLowerCase();
+        return this.options.filter(option => option.toLowerCase().includes(filterValue));
+    }
+
+    public getGoods(): void {
+        this.cartService.getGoods().subscribe(goods => this.goods = goods);
+    }
+
+    public getSum(): void {
+        this.subTotalSum = +this.cartService.getTotalSum(this.goods).toFixed(2);
+        this.taxSum = +((this.tax * this.subTotalSum) / 100).toFixed(2);
+        this.totalSum = +(this.subTotalSum + (this.tax * this.subTotalSum) / 100).toFixed(2);
+    }
+
+    public applyPromo(): void {
+        const control = this.checkoutForm.get('promo') as FormArray;
+        if (control.value) {
+            this.totalSum = +(this.totalSum - (this.totalSum * this.promoPercentages / 100)).toFixed(2);
+        }
+    }
+
+    public removeGood(id: number): void {
+        this.cartService.removeGood(id).subscribe(goods => this.goods = goods);
+        this.patchGoods();
+        this.getSum();
+    }
+
+    public handleCountValue(event: Event, good: GoodInterface): void {
+        const target = event.target as HTMLInputElement;
+        if (+target.value < 1) {
+            target.value = '1';
```

Comment:
Is it possible to rewrite value from `event` data?

Reply:
Use `valueNumber` instead of just `value`.

Reply:
Possible, it works